### PR TITLE
Issue #7 feat: SMS OTP Reset flow provider

### PR DESCRIPTION
**Pull request by shriharshs**
PR for SMS OTP Authentication of Forgot password reset credentials flow

### CODE REVIEW

File: provider/EmailAndPhone.iml [hunk 1]
```diff
@@ -0,0 +1,61 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (kochhar) [inline provider/EmailAndPhone.iml]: @shriharshs are you sure this file is needed? seems like an IDE file which may not need to be committed.

File: provider/keycloak-email-phone-autthenticator.iml [hunk 2]
```diff
@@ -0,0 +1,60 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (kochhar) [inline provider/keycloak-email-phone-autthenticator.iml]: @shriharshs Same note for all .iml files

File: provider/pom.xml [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+        </dependency>
+        <dependency>
+            <groupId>com.amazonaws</groupId>
+            <artifactId>aws-java-sdk-sns</artifactId>
+            <version>1.11.173</version>
+        </dependency>
+    </dependencies>
+
+
+</project>
```
Comment (kochhar) [inline provider/pom.xml]: @shriharshs new line at end of file

File: provider/src/main/java/org/sunbird/aws/snsclient/SnsClientFactory.java [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+import com.amazonaws.auth.BasicAWSCredentials;
+import com.amazonaws.regions.Region;
+import com.amazonaws.regions.Regions;
+import com.amazonaws.services.sns.AmazonSNSClient;
+import com.amazonaws.services.sns.model.SetSMSAttributesRequest;
+
+/**
+ * Created by nickpack on 09/08/2017.
+ */
+public class SnsClientFactory {
```
Comment (kochhar) [inline provider/src/main/java/org/sunbird/aws/snsclient/SnsClientFactory.java]: @shriharshs I suggest that this factory implements a consistent interface which can be implemented by other factories which provide SMS sending channels.

cc:@coolbung

File: provider/src/main/java/org/sunbird/aws/snsclient/SnsNotificationService.java [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+import com.amazonaws.services.sns.model.PublishRequest;
+import com.amazonaws.services.sns.model.PublishResult;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Created by nickpack on 09/08/2017.
+ */
+public class SnsNotificationService {
```
Comment (kochhar) [inline provider/src/main/java/org/sunbird/aws/snsclient/SnsNotificationService.java]: @shriharshs This class should also implement a common interface for SMS sending channels.

File: provider/src/main/java/org/sunbird/keycloak/KeycloakSmsAuthenticator.java [hunk 6]
```diff
@@ -0,0 +1,379 @@
...
+        EXPIRED
+    }
+
+
+    public void authenticate(AuthenticationFlowContext context) {
+        logger.debug("authenticate called ... context = " + context);
+        UserModel user = context.getUser();
+        AuthenticatorConfigModel config = context.getAuthenticatorConfig();
+
+        List<String> mobileNumberCreds = user.getAttribute("mobile");
```
Comment (kochhar) [inline provider/src/main/java/org/sunbird/keycloak/KeycloakSmsAuthenticator.java]: @shriharshs the attribute which stores the mobile number should come from configuration. Each sunbird instance may choose a different name for this attribute.

**Comment by shriharshs** [comment]
@coolbung - I think its better be out, because it might confuse with the themes folder of keycloak, and this folder you can easily be imported as a project in IDE.

**Comment by coolbung** [comment]
I was thinking to mimic the folder structure when keycloak gets deployed. Or is that not useful ?

cc @kochhar

File: provider/src/main/java/org/sunbird/keycloak/KeycloakSmsAuthenticator.java [hunk 7]
```diff
@@ -0,0 +1,379 @@
...
+                    logger.debug("Code is expired !!");
+                    result = CODE_STATUS.EXPIRED;
+                }
+            }*/
+        }
+        logger.debug("result : " + result);
+        return result;
+    }
+
+//    public void sendPasswordResetLink (String link, long expirationInMinutes, AuthenticationFlowContext context) throws EmailException {
```
Comment (kochhar) [inline provider/src/main/java/org/sunbird/keycloak/KeycloakSmsAuthenticator.java]: @shriharshs please delete commented code.

File: provider/src/main/java/org/sunbird/keycloak/KeycloakSmsAuthenticator.java [hunk 8]
```diff
@@ -0,0 +1,379 @@
...
+
+        if (mobileNumberCreds != null && !mobileNumberCreds.isEmpty()) {
+            mobileNumber = mobileNumberCreds.get(0);
+        }
+
+        if (!TextUtils.isEmpty(userEmail)) {
+            logger.debug("Email is not null - " + userEmail);
+        }
+
+        if (mobileNumber != null) {
```
Comment (kochhar) [inline provider/src/main/java/org/sunbird/keycloak/KeycloakSmsAuthenticator.java]: @shriharshs suggest moving the sms sending to a separate method which can be tested independently.

File: provider/src/main/java/org/sunbird/keycloak/KeycloakSmsAuthenticator.java [hunk 9]
```diff
@@ -0,0 +1,379 @@
+package org.sunbird.keycloak;
```
Comment (kochhar) [inline provider/src/main/java/org/sunbird/keycloak/KeycloakSmsAuthenticator.java]: @shriharshs I suggest adding test cases for this class. Please share a list of test cases which you believe will be relevant to test so we can prioritise accordingly.

File: provider/src/main/java/org/sunbird/keycloak/KeycloakSmsAuthenticatorCredentialProvider.java [hunk 10]
```diff
@@ -0,0 +1,110 @@
+package org.sunbird.keycloak;
```
Comment (kochhar) [inline provider/src/main/java/org/sunbird/keycloak/KeycloakSmsAuthenticatorCredentialProvider.java]: @shriharshs This class will also need some test cases.

File: provider/src/main/java/org/sunbird/keycloak/KeycloakSmsAuthenticatorUtil.java [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+        String text = KeycloakSmsAuthenticatorUtil.getConfigString(config, KeycloakSmsAuthenticatorConstants.CONF_PRP_SMS_TEXT);
+        text = text.replaceAll("%sms-code%", code);
+        text = text.replaceAll("%phonenumber%", mobileNumber);
+
+        return text;
+    }
+
+    public static String setDefaultCountryCodeIfZero(String mobileNumber) {
+        if (mobileNumber.startsWith("0")) {
+            mobileNumber = "+91" + mobileNumber.substring(1);
```
Comment (kochhar) [inline provider/src/main/java/org/sunbird/keycloak/KeycloakSmsAuthenticatorUtil.java]: @shriharshs 
- Please make default country code a config option
- Please make initial digit check configurable

**Comment by kochhar** [comment]
@coolbung @shriharshs would be good to put this inside a keycloak folder. There may be other auth utilities unrelated to keycloak in this repo.

/keycloak/sms-provider/src/... can be the directory imported into an IDE.

File: provider/EmailAndPhone.iml (hunk 1 above)
Comment (shriharshs) [reply provider/EmailAndPhone.iml]: @kochhar Sometimes IDE throws error without this .iml file when you import the code. We faced this issue in Android as well, so thought I'll add it.

**Comment by shriharshs** [comment]
@kochhar Other than test cases, other changes are done! :)

**Comment by shriharshs** [comment]
@kochhar Created test case stubs.

File: keycloak/sms-provider/src/main/java/org/sunbird/keycloak/KeycloakSmsAuthenticatorConstants.java [hunk 12]
```diff
@@ -6,6 +6,8 @@
 public class KeycloakSmsAuthenticatorConstants {
     public static final String ATTR_MOBILE = "mobile";
     public static final String ANSW_SMS_CODE = "smsCode";
+    public static final String COUNTRY_CODE = "+91";
```
Comment (kochhar) [inline keycloak/sms-provider/src/main/java/org/sunbird/keycloak/KeycloakSmsAuthenticatorConstants.java]: This must be configuration not code. Please add a TODO.

File: keycloak/sms-provider/src/main/java/org/sunbird/keycloak/KeycloakSmsAuthenticatorConstants.java [hunk 13]
```diff
@@ -6,6 +6,8 @@
 public class KeycloakSmsAuthenticatorConstants {
     public static final String ATTR_MOBILE = "mobile";
```
Comment (kochhar) [inline keycloak/sms-provider/src/main/java/org/sunbird/keycloak/KeycloakSmsAuthenticatorConstants.java]: This must be configuration not code.

File: keycloak/sms-provider/src/main/java/org/sunbird/keycloak/KeycloakSmsAuthenticatorConstants.java [hunk 14]
```diff
@@ -6,6 +6,8 @@
...
     public static final String ANSW_SMS_CODE = "smsCode";
+    public static final String COUNTRY_CODE = "+91";
+    public static final String DEFAULT_COUNTRY_CODE = "0";
```
Comment (kochhar) [inline keycloak/sms-provider/src/main/java/org/sunbird/keycloak/KeycloakSmsAuthenticatorConstants.java]: Also here code -> config.

File: keycloak/sms-provider/src/main/java/org/sunbird/keycloak/KeycloakSmsAuthenticatorConstants.java (hunk 12 above)
Comment (shriharshs) [reply keycloak/sms-provider/src/main/java/org/sunbird/keycloak/KeycloakSmsAuthenticatorConstants.java]: @kochhar Ok, I'll add that to the attributes of the configurations. But what if user does not set these? Should we have them as default ones?

**Pull request merged**